### PR TITLE
[DCP] Add DefaultStager example to distributed async checkpoint recipe

### LINKED ISSUE #3710: [DCP] Add DefaultStager example to distributed async checkpoint recipe
### 🚀 Feature Request

**Description**
The current `distributed_async_checkpoint_recipe` covers basic usage of `dcp.async_save` and Pinned Memory optimization. However, it does not cover the **fully asynchronous staging** capabilities introduced in PyTorch 2.9 via `DefaultStager`.

Even with `async_save`, the Device-to-Host (D2H) copy (staging phase) typically happens on the main thread, which can block the training loop.

**Proposal**
I would like to update the tutorial to include a new section on **"Fully Asynchronous Staging with DefaultStager"**.

This update will demonstrate:
1.  How to use the `async_stager=DefaultStager()` argument.
2.  How to correctly synchronize staging to achieve full overlap between the D2H copy and the **Forward + Backward** pass of the next step.
3.  Timeline comparison between standard async save and stager-based async save.

I have already prepared the content and code example.

### PR DESCRIPTION
**Pull request by niyunsheng**
Fixes #3710 

## Description

This PR updates the `distributed_async_checkpoint_recipe` to include the `DefaultStager` functionality introduced in PyTorch 2.9.

**Motivation:**
In large-scale training, even with standard `async_save`, the initial memory copy (Staging phase, GPU -> CPU) occurs on the main thread. This blocks the training loop. This PR introduces `DefaultStager`, which offloads this copy to a background thread, enabling **full computation-communication overlap**.

**Key Changes:**
1.  **New Section**: Added "Fully Asynchronous Staging with DefaultStager" to the recipe.
2.  **Version Note**: Added `.. versionadded:: 2.9` to indicate version requirements.
3.  **Advanced Example**: Provided a code example demonstrating how to overlap the D2H copy with the **entire Forward and Backward pass**:
    * Check `staging_completion` *after* backward but *before* `optimizer.step()` to ensure data consistency while maximizing parallel execution.
    * Check `upload_completion` before the next save to manage memory backpressure.
4.  **Authors**: Added myself to the author list.

## Checklist

- [x] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [x] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

cc @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0 @haochengsong @Saiteja64

### CODE REVIEW

**Comment by niyunsheng** [comment]
Hi @LucasLLC, gentle ping on this.

Just to clarify context: This tutorial covers an existing feature (DefaultStager) and is independent of my other open PRs. It does not require any upstream changes.

It has been passing CI for two weeks. Could you please take a look when you have a moment? Thanks!

**Comment by AlannaBurke** [comment]
Preview link: https://docs-preview.pytorch.org/pytorch/tutorials/3711/recipes/distributed_async_checkpoint_recipe.html

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 1]
```diff
@@ -1,7 +1,7 @@
 Asynchronous Saving with Distributed Checkpoint (DCP)
 =====================================================
 
-**Author:** `Lucas Pasqualin <https://github.com/lucasllc>`__, `Iris Zhang <https://github.com/wz337>`__, `Rodrigo Kumpera <https://github.com/kumpera>`__, `Chien-Chin Huang <https://github.com/fegin>`__
+**Author:** `Lucas Pasqualin <https://github.com/lucasllc>`__, `Iris Zhang <https://github.com/wz337>`__, `Rodrigo Kumpera <https://github.com/kumpera>`__, `Chien-Chin Huang <https://github.com/fegin>`__, `Yunsheng Ni <https://github.com/niyunsheng>`__
 
 Checkpointing is often a bottle-neck in the critical path for distributed training workloads, incurring larger and larger costs as both model and world sizes grow.
```
Comment (AlannaBurke) [inline recipes_source/distributed_async_checkpoint_recipe.rst]:
```suggestion
Checkpointing is often a bottleneck in the critical path for distributed training workloads, incurring larger and larger costs as both model and world sizes grow.
```

File: recipes_source/distributed_async_checkpoint_recipe.rst
Comment (AlannaBurke) [inline recipes_source/distributed_async_checkpoint_recipe.rst]:
```suggestion
Before getting started with Asynchronous Checkpointing, it's important to understand its differences and limitations as compared to synchronous checkpointing.
```

Comment (AlannaBurke) [inline recipes_source/distributed_async_checkpoint_recipe.rst]:
```suggestion
* Checkpoint Management - Since checkpointing is asynchronous, it is up to the user to manage concurrently run checkpoints. 
    In general, users can
```

Comment (AlannaBurke) [inline recipes_source/distributed_async_checkpoint_recipe.rst]:
```suggestion
employ their own management strategies by handling the future object returned form ``async_save``. For most users, we recommend limiting
```

Comment (AlannaBurke) [inline recipes_source/distributed_async_checkpoint_recipe.rst]:
```suggestion
            # this line automatically manages FSDP FQNs, as well as sets the default state dict type to FSDP.SHARDED_STATE_DICT
```

File: recipes_source/distributed_async_checkpoint_recipe.rst [hunk 2]
```diff
@@ -279,6 +279,154 @@ checkpoint requests users can take advantage of direct memory access to speed up
...
+        Note: We take advantage of this wrapper to hande calling distributed state dict methods on the model
+        and optimizer.
+        """
+
+        def __init__(self, model, optimizer=None):
+            self.model = model
+            self.optimizer = optimizer
+
+        def state_dict(self):
+            # this line automatically manages FSDP FQN's, as well as sets the default state dict type to FSDP.SHARDED_STATE_DICT
```
Comment (AlannaBurke) [inline recipes_source/distributed_async_checkpoint_recipe.rst]:
```suggestion
            # this line automatically manages FSDP FQNs, as well as sets the default state dict type to FSDP.SHARDED_STATE_DICT
```

**Review by AlannaBurke — CHANGES_REQUESTED** [review]
Just a few grammar and minor changes. Otherwise this looks good to me.

**Comment by niyunsheng** [comment]
Hi @AlannaBurke , thanks for the review!

I've applied your grammar and other suggestions.

Extra Fix: While previewing the file locally (`make html-noplot`), I noticed another small display issue in this file that was missing from the review, so I went ahead and fixed that as well.

The file should be perfect now. Ready for another look!

**Comment by niyunsheng** [comment]
Thanks @LucasLLC for the approval!

@AlannaBurke I've pushed a fix for the remaining lint/whitespace errors. Since I'm an external contributor, could you please approve the CI workflows to run?

Once the checks are green, this should be ready to land. Thanks!

**Comment by niyunsheng** [comment]
"Hi @AlannaBurke @LucasLLC,

I've just synced this branch with main to ensure everything is up-to-date.

Could you please help approve the CI workflows to run? Once the checks pass, this should be ready to merge. Thanks!

**Review by AlannaBurke — APPROVED** [review]

**Comment by niyunsheng** [comment]
Hi @AlannaBurke,

Just a gentle bump on this. Since this PR has been open for a month and is now approved (thanks!), I'd really love to get it across the finish line.

The CI workflows are currently stuck in "awaiting approval" after the recent sync with main. Could you please help trigger them so we can merge?

Thanks!

**Comment by niyunsheng** [comment]
Hi @AlannaBurke and @svekars,

I've just updated the branch to resolve the out-of-date status.

Since this PR has been approved for two weeks, I'd really appreciate it if either of you could help trigger the final CI.

With the Lunar New Year holiday coming up in two weeks, I'd love to see this land beforehand so we don't carry it over.

Thanks for your help!
